lib/Transforms/Instrumentation/BranchSentinelPass.cc

```diff
@@ -0,0 +1,92 @@
+/**
+ * This pass adds a branch sentinel call instruction(intrinsic) before
+ * every conditional branch.
+ */
+#include "seahorn/Analysis/SeaBuiltinsInfo.hh"
+#include "seahorn/InitializePasses.hh"
+#include "seahorn/Support/SeaDebug.h"
+#include "seahorn/Support/SeaLog.hh"
+#include "seahorn/config.h"
+
+#include "llvm/ADT/Statistic.h"
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instruction.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Pass.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+
+using namespace llvm;
+
+// TODO: add stats if useful
+
+namespace {
+struct AddBranchSentinelPass : public FunctionPass {
+  static char ID;
+  seahorn::SeaBuiltinsInfo *m_SBI;
+
+  AddBranchSentinelPass() : FunctionPass(ID) {}
+  void augmentBranchInstWithSentinel(BranchInst &BI, IRBuilder<> &B);
+  bool runOnFunction(Function &F) override;
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.addRequired<seahorn::SeaBuiltinsInfoWrapperPass>();
+    // preserve nothing
```

Comment:
should probably preserve everything. Otherwise hard to combine with other passes